### PR TITLE
Add: interface of Metadata report

### PR DESCRIPTION
**Pull request by hxmhlt**
**What this PR does**:
define interface of Metadata report 
**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

File: common/extension/metadata_report_factory.go [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+ */
+
+package extension
+
+import (
+	"github.com/apache/dubbo-go/metadata"
+)
+
+var (
+	metaDataReportFactories = make(map[string]func() metadata.MetadataReportFactory)
```
Comment (AlexStocks) [inline common/extension/metadata_report_factory.go]: set its init length to 8.

File: common/extension/metadata_report_factory.go [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+)
+
+// SetMetadataReportFactory ...
```
Comment (AlexStocks) [inline common/extension/metadata_report_factory.go]: add remark.

File: common/extension/metadata_report_factory.go (hunk 1 above)
Comment (AlexStocks) [reply common/extension/metadata_report_factory.go]: secendly, do u think it should be protected by a lock?

File: common/extension/metadata_report_factory.go [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+func SetMetadataReportFactory(name string, v func() metadata.MetadataReportFactory) {
+	metaDataReportFactories[name] = v
+}
+
+// GetConfigCenterFactory ...
```
Comment (AlexStocks) [inline common/extension/metadata_report_factory.go]: add remark.

File: common/extension/metadata_report_factory.go [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+func GetMetadataReportFactory(name string) metadata.MetadataReportFactory {
+	if metaDataReportFactories[name] == nil {
+		panic("metadata report for " + name + " is not existing, make sure you have import the package.")
```
Comment (AlexStocks) [inline common/extension/metadata_report_factory.go]: panic? maybe u should return a error.

File: config/metadata_report_config.go [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package config
+
+import (
+	"github.com/apache/dubbo-go/metadata"
```
Comment (AlexStocks) [inline config/metadata_report_config.go]: move it to the third import block.

File: config/metadata_report_config.go [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+
+// Prefix ...
+func (c *MetadataReportConfig) Prefix() string {
+	return constant.MetadataReportPrefix
+}
+
+// UnmarshalYAML ...
+func (c *MetadataReportConfig) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	if err := defaults.Set(c); err != nil {
+		return err
```
Comment (AlexStocks) [inline config/metadata_report_config.go]: perros.WithStack(err)

File: config/metadata_report_config.go [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+	}
+	type plain MetadataReportConfig
+	if err := unmarshal((*plain)(c)); err != nil {
+		return err
```
Comment (AlexStocks) [inline config/metadata_report_config.go]: perros.WithStack(err)

File: config/metadata_report_config.go [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+	url.SetParam("metadata", url.Protocol)
+	return &url, nil
+}
+
+// StartMetadataReport: The entry of metadata report start
+func startMetadataReport() error {
+	metadataConfig := consumerConfig.ApplicationConfig.MetadataType
+	if consumerConfig.MetadataReportConfig == nil {
+		return nil
+	} else {
```
Comment (AlexStocks) [inline config/metadata_report_config.go]: delete the else.

File: config/metadata_report_config.go [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+	)
+	if err != nil {
+		return nil, perrors.New("Invalid MetadataReportConfig.")
+	}
+	url.SetParam("metadata", url.Protocol)
+	return &url, nil
+}
+
+// StartMetadataReport: The entry of metadata report start
+func startMetadataReport() error {
```
Comment (AlexStocks) [inline config/metadata_report_config.go]: in this func, there are so many if-else. maybe u should use switch instead.

File: config/provider_config.go [hunk 10]
```diff
@@ -103,7 +103,10 @@ func ProviderInit(confProFile string) error {
 			n.InterfaceId = k
 		}
 	}
-
+	//start the metadata report if config set
+	if err := startMetadataReport(); err != nil {
+		return perrors.WithMessagef(err, "Provider start metadata report error ,error is {%#v}", err)
```
Comment (AlexStocks) [inline config/provider_config.go]: Provider starts  to report metadata error,  and the error is

File: config/metadata_report_config.go [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+		return err
+	}
+	type plain MetadataReportConfig
+	if err := unmarshal((*plain)(c)); err != nil {
+		return err
+	}
+	return nil
+}
+
+// ToUrl ...
```
Comment (AlexStocks) [inline config/metadata_report_config.go]: add remark.

File: config/metadata_report_config.go [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+	Protocol   string            `required:"true"  yaml:"protocol"  json:"protocol,omitempty"`
+	Address    string            `yaml:"address" json:"address,omitempty" property:"address"`
+	Username   string            `yaml:"username" json:"username,omitempty" property:"username"`
+	Password   string            `yaml:"password" json:"password,omitempty"  property:"password"`
+	Params     map[string]string `yaml:"params" json:"params,omitempty" property:"params"`
+	TimeoutStr string            `yaml:"timeout" default:"5s" json:"timeout,omitempty" property:"timeout"` // unit: second
+	Group      string            `yaml:"group" json:"group,omitempty" property:"group"`
+}
+
+// Prefix ...
```
Comment (AlexStocks) [inline config/metadata_report_config.go]: add remark.

File: config/metadata_report_config.go [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+func (c *MetadataReportConfig) Prefix() string {
+	return constant.MetadataReportPrefix
+}
+
+// UnmarshalYAML ...
```
Comment (AlexStocks) [inline config/metadata_report_config.go]: add remark

File: config/metadata_report_config.go [hunk 14]
```diff
@@ -0,0 +1,102 @@
...
+	"github.com/creasty/defaults"
+	perrors "github.com/pkg/errors"
+)
+
+import (
+	"github.com/apache/dubbo-go/common"
+	"github.com/apache/dubbo-go/common/constant"
+)
+
+// MethodConfig ...
```
Comment (AlexStocks) [inline config/metadata_report_config.go]: add remark

File: metadata/definition/definition.go [hunk 15]
```diff
@@ -0,0 +1,41 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package definition
+
+type ServiceDefinition struct {
```
Comment (AlexStocks) [inline metadata/definition/definition.go]: add remark.

File: metadata/definition/definition.go [hunk 16]
```diff
@@ -0,0 +1,41 @@
...
+	Methods       []MethodDefinition
+	Types         []TypeDefinition
+}
+
+type MethodDefinition struct {
+	Name           string
+	ParameterTypes []string
+	ReturnType     string
+	Parameters     []TypeDefinition
+}
```
Comment (AlexStocks) [inline metadata/definition/definition.go]: add remark.

File: metadata/definition/definition.go [hunk 17]
```diff
@@ -0,0 +1,41 @@
...
+
+package definition
+
+type ServiceDefinition struct {
+	CanonicalName string
+	CodeSource    string
+	Methods       []MethodDefinition
+	Types         []TypeDefinition
+}
+
```
Comment (AlexStocks) [inline metadata/definition/definition.go]: add remark.

File: metadata/exporter.go [hunk 18]
```diff
@@ -0,0 +1,27 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package metadata
+
+import "github.com/apache/dubbo-go/common"
+
+type MetadataExporter interface {
```
Comment (AlexStocks) [inline metadata/exporter.go]: add remark.

File: metadata/exporter.go [hunk 19]
```diff
@@ -0,0 +1,27 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package metadata
+
+import "github.com/apache/dubbo-go/common"
```
Comment (AlexStocks) [inline metadata/exporter.go]: import ()

File: metadata/identifier/base_metadata_identifier.go [hunk 20]
```diff
@@ -0,0 +1,92 @@
...
+
+import (
+	"encoding/base64"
+)
+
+import (
+	"github.com/apache/dubbo-go/common/constant"
+)
+
+type BaseMetadataIdentifier interface {
```
Comment (AlexStocks) [inline metadata/identifier/base_metadata_identifier.go]: add remark.

File: metadata/identifier/base_metadata_identifier.go [hunk 21]
```diff
@@ -0,0 +1,92 @@
...
+	getFilePathKey(params ...string) string
+	getIdentifierKey(params ...string) string
+}
+
+type BaseServiceMetadataIdentifier struct {
```
Comment (AlexStocks) [inline metadata/identifier/base_metadata_identifier.go]: add remark.

File: metadata/identifier/base_metadata_identifier.go [hunk 22]
```diff
@@ -0,0 +1,92 @@
...
+	serviceInterface string
+	version          string
+	group            string
+	side             string
+}
+
+// joinParams...
```
Comment (AlexStocks) [inline metadata/identifier/base_metadata_identifier.go]: delete the unuseful remark.

File: metadata/identifier/base_metadata_identifier.go [hunk 23]
```diff
@@ -0,0 +1,92 @@
...
+func joinParams(joinChar string, params []string) string {
+	var joinedStr string
+	for _, param := range params {
+		joinedStr += joinChar
+		joinedStr += param
+	}
+	return joinedStr
+}
+
+// getIdentifierKey...
```
Comment (AlexStocks) [inline metadata/identifier/base_metadata_identifier.go]: delete the unuseful remark.

File: metadata/identifier/base_metadata_identifier.go [hunk 24]
```diff
@@ -0,0 +1,92 @@
...
+func (mdi *BaseServiceMetadataIdentifier) getIdentifierKey(params ...string) string {
+	return mdi.serviceInterface +
+		constant.KEY_SEPARATOR + mdi.version +
+		constant.KEY_SEPARATOR + mdi.group +
+		constant.KEY_SEPARATOR + mdi.side +
+		joinParams(constant.KEY_SEPARATOR, params)
+}
+
+// getIdentifierKey...
```
Comment (AlexStocks) [inline metadata/identifier/base_metadata_identifier.go]: delete the unuseful remark.

File: metadata/identifier/base_metadata_identifier.go [hunk 25]
```diff
@@ -0,0 +1,92 @@
...
+		joinParams(constant.PATH_SEPARATOR, params)
+
+}
+
+// serviceToPath...
+func serviceToPath(serviceInterface string) string {
+	if serviceInterface == constant.ANY_VALUE {
+		return ""
+	} else {
+		decoded, err := base64.URLEncoding.DecodeString(serviceInterface)
```
Comment (AlexStocks) [inline metadata/identifier/base_metadata_identifier.go]: delete the unuseful remark.

File: metadata/identifier/base_metadata_identifier.go [hunk 26]
```diff
@@ -0,0 +1,92 @@
...
+		if err != nil {
+			return ""
+		}
+		return string(decoded)
+	}
+
+}
+
+//withPathSeparator...
```
Comment (AlexStocks) [inline metadata/identifier/base_metadata_identifier.go]: delete the unuseful remark.

File: metadata/identifier/metadata_identifier.go [hunk 27]
```diff
@@ -0,0 +1,33 @@
...
+ */
+
+package identifier
+
+type MetadataIdentifier struct {
+	application string
+	BaseMetadataIdentifier
+}
+
+// getIdentifierKey...
```
Comment (AlexStocks) [inline metadata/identifier/metadata_identifier.go]: delete the unuseful remark.

File: metadata/identifier/metadata_identifier.go [hunk 28]
```diff
@@ -0,0 +1,33 @@
...
+func (mdi *MetadataIdentifier) getIdentifierKey(params ...string) string {
+	return mdi.BaseMetadataIdentifier.getIdentifierKey(mdi.application)
+}
+
+// getIdentifierKey...
```
Comment (AlexStocks) [inline metadata/identifier/metadata_identifier.go]: delete the unuseful remark.

File: metadata/identifier/service_metadata_identifier.go [hunk 29]
```diff
@@ -0,0 +1,36 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package identifier
+
+import "github.com/apache/dubbo-go/common/constant"
```
Comment (AlexStocks) [inline metadata/identifier/service_metadata_identifier.go]: import ()

File: metadata/identifier/service_metadata_identifier.go [hunk 30]
```diff
@@ -0,0 +1,36 @@
...
+
+type ServiceMetadataIdentifier struct {
```
Comment (AlexStocks) [inline metadata/identifier/service_metadata_identifier.go]: add remark for this struct

File: metadata/identifier/service_metadata_identifier.go [hunk 31]
```diff
@@ -0,0 +1,36 @@
...
+	protocol string
+	BaseMetadataIdentifier
+}
+
+// getIdentifierKey...
+func (mdi *ServiceMetadataIdentifier) getIdentifierKey(params ...string) string {
+	return mdi.BaseMetadataIdentifier.getIdentifierKey(mdi.protocol + constant.KEY_REVISON_PREFIX + mdi.revision)
+}
+
+// getIdentifierKey...
```
Comment (AlexStocks) [inline metadata/identifier/service_metadata_identifier.go]: delete the unuseful remark or add useful remark.

File: metadata/identifier/service_metadata_identifier.go [hunk 32]
```diff
@@ -0,0 +1,36 @@
...
+
+import "github.com/apache/dubbo-go/common/constant"
+
+type ServiceMetadataIdentifier struct {
+	revision string
+	protocol string
+	BaseMetadataIdentifier
+}
+
+// getIdentifierKey...
```
Comment (AlexStocks) [inline metadata/identifier/service_metadata_identifier.go]: delete the unuseful remark or add useful remark.

File: metadata/identifier/subscribe_metadata_identifier.go [hunk 33]
```diff
@@ -0,0 +1,16 @@
...
+	revision string
+	BaseMetadataIdentifier
+}
+
+// getIdentifierKey...
+func (mdi *SubscriberMetadataIdentifier) getIdentifierKey(params ...string) string {
+	return mdi.BaseMetadataIdentifier.getIdentifierKey(mdi.revision)
+}
+
+// getIdentifierKey...
```
Comment (AlexStocks) [inline metadata/identifier/subscribe_metadata_identifier.go]: delete the unuseful remark or add useful remark.

File: metadata/identifier/subscribe_metadata_identifier.go [hunk 34]
```diff
@@ -0,0 +1,16 @@
+package identifier
+
+type SubscriberMetadataIdentifier struct {
+	revision string
+	BaseMetadataIdentifier
+}
+
+// getIdentifierKey...
```
Comment (AlexStocks) [inline metadata/identifier/subscribe_metadata_identifier.go]: delete the unuseful remark or add useful remark.

File: metadata/identifier/subscribe_metadata_identifier.go [hunk 35]
```diff
@@ -0,0 +1,16 @@
+package identifier
+
```
Comment (AlexStocks) [inline metadata/identifier/subscribe_metadata_identifier.go]: add remark.

File: metadata/instance.go [hunk 36]
```diff
@@ -0,0 +1,40 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package metadata
+
+import (
```
Comment (AlexStocks) [inline metadata/instance.go]: split it.

File: metadata/instance.go [hunk 37]
```diff
@@ -0,0 +1,40 @@
...
+	"github.com/apache/dubbo-go/common/extension"
+	"sync"
+)
+
+var (
+	instance MetadataReport
+	once     sync.Once
+)
+
+// GetEnvInstance ...
```
Comment (AlexStocks) [inline metadata/instance.go]: delete the unuseful remark or add useful remark.

File: metadata/instance.go [hunk 38]
```diff
@@ -0,0 +1,40 @@
...
+func Init(url *common.URL) {
+	once.Do(func() {
+		instance = extension.GetMetadataReportFactory(url.Protocol).createMetadataReport(url)
+	})
+}
+
+func GetMetadataReportInstance() MetadataReport {
```
Comment (AlexStocks) [inline metadata/instance.go]: add useful remark.

File: metadata/report_factory.go [hunk 39]
```diff
@@ -0,0 +1,28 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package metadata
+
+import "github.com/apache/dubbo-go/common"
```
Comment (AlexStocks) [inline metadata/report_factory.go]: import ()

File: metadata/report_factory.go [hunk 40]
```diff
@@ -0,0 +1,28 @@
...
+
+var (
+	MetadataReportInstance MetadataReport
+)
+
```
Comment (AlexStocks) [inline metadata/report_factory.go]: add useful remark.

**Comment by AlexStocks** [comment]
@hxmhlt pls add some words of "What this PR does:" for this pr.

File: common/extension/metadata_report_factory.go (hunk 1 above)
Comment (hxmhlt) [reply common/extension/metadata_report_factory.go]: extension map do not need to lock

File: common/extension/metadata_report_factory.go (hunk 4 above)
Comment (hxmhlt) [reply common/extension/metadata_report_factory.go]: Extension is designed for dubbogo plugin  system, we can not start the program when user's plugin is not exist.

File: config/metadata_report_config.go (hunk 9 above)
Comment (hxmhlt) [reply config/metadata_report_config.go]: The case is not for one variable.So switch may not work.

**Pull request merged**